### PR TITLE
Warn about tech preview of checksums.

### PR DESCRIPTION
**Pull request by jessfraz**
ping @icecrime @crosbymichael @shykes @tianon

### CODE REVIEW

File: graph/pull.go [hunk 1]
```diff
@@ -491,7 +491,7 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
 
 				io.Copy(tmpFile, utils.ProgressReader(ioutil.NopCloser(tarSumReader), int(l), out, sf, false, utils.TruncateID(img.ID), "Downloading"))
 
-				out.Write(sf.FormatProgress(utils.TruncateID(img.ID), "Verifying Checksum", nil))
+				out.Write(sf.FormatProgress(utils.TruncateID(img.ID), "Verifying Checksum, this is a tech preview, don't rely on it for security yet.", nil))
```
Comment (tiborvass) [inline graph/pull.go]: maybe replace the first comma `,` by ` - ` ? other than that, looks good!

Comment (jfrazelle) [inline graph/pull.go]: cool will do!

File: graph/pull.go [hunk 2]
```diff
@@ -428,7 +428,7 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
 	}
 
 	if verified {
-		out.Write(sf.FormatStatus(repoInfo.CanonicalName+":"+tag, "The image you are pulling has been verified"))
+		out.Write(sf.FormatStatus(repoInfo.CanonicalName+":"+tag, "The image you are pulling has been verified.\nThis is a tech preview, don't rely on it for security yet."))
```
Comment (jfrazelle) [inline graph/pull.go]: I actually think we should just remove this one, it is only verifying the manifest not the layers.

Comment (tianon) [inline graph/pull.go]: +1

File: graph/pull.go [hunk 3]
```diff
@@ -428,10 +428,10 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
 	}
 
 	if verified {
-		out.Write(sf.FormatStatus(repoInfo.CanonicalName+":"+tag, "The image you are pulling has been verified"))
-	} else {
-		out.Write(sf.FormatStatus(tag, "Pulling from %s", repoInfo.CanonicalName))
+		log.Printf("Image manifest verified for %s:%s", repoInfo.CanonicalName, tag)
 	}
+	out.Write(sf.FormatStatus(tag, "Pulling from %s", repoInfo.CanonicalName))
```
Comment (icecrime) [inline graph/pull.go]: Shouldn't this happen *before* the `"image manifest verified"` log? I just find the sequence a bit weird:

```
Image manifest verified for foo/bar:baz
Pulling from foo/bar
```

Comment (tiborvass) [inline graph/pull.go]: @icecrime I read it as "It's pulling the image", not the "manifest". So it verifies the manifest first and then pulls the image data. Should we clarify it somehow?

Comment (jfrazelle) [inline graph/pull.go]: The log will go to the Daemon logs anyways

Comment (jfrazelle) [inline graph/pull.go]: do you think this is ok or should I still move it?

Comment (tiborvass) [inline graph/pull.go]: I'm personally okay with this, but what about the others?

Comment (tianon) [inline graph/pull.go]: If it's just in the daemon log, that doesn't bother me. :)

(Nobody reads the daemon logs...)

Comment (icecrime) [inline graph/pull.go]: Yes sorry I didn't notice this was going to the logs anyways. I'm okay with this.

**Comment by icecrime** [comment]
Ping @dmp42 @dmcgowan: I think you guys should give your go on this.

File: graph/pull.go [hunk 4]
```diff
@@ -491,7 +491,7 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
 
 				io.Copy(tmpFile, utils.ProgressReader(ioutil.NopCloser(tarSumReader), int(l), out, sf, false, utils.TruncateID(img.ID), "Downloading"))
 
-				out.Write(sf.FormatProgress(utils.TruncateID(img.ID), "Verifying Checksum", nil))
+				out.Write(sf.FormatProgress(utils.TruncateID(img.ID), "Verifying Checksum - this is a tech preview, don't rely on it for security yet.", nil))
```
Comment (dmcgowan) [inline graph/pull.go]: This will only flash for a fraction of a second in most cases

Comment (jfrazelle) [inline graph/pull.go]: well I still think it needs to be said, do you suggest putting it somewhere else so it stays

Comment (dmcgowan) [inline graph/pull.go]: I think the only other place it belongs is next to the verified message.  Keep in mind the current verified message will only show up for official images signed by official image build key.  While we might change the checksum method, the fact that we are verifying the checksum using tarsum IMO does not need to need the preview subtext.

Comment (jfrazelle) [inline graph/pull.go]: I think it does need the preview subtext and I believe @shykes thought so
as well unless I misunderstood.

Comment (dmcgowan) [inline graph/pull.go]: I may have misunderstood as well, that the subtext belongs with the "verification" claim not the checksum claim.

**Comment by jfrazelle** [comment]
@dmcgowan @icecrime let me know about the latest changes

**Comment by jfrazelle** [comment]
added a test

**Comment by dmcgowan** [comment]
I like it, LGTM

**Comment by icecrime** [comment]
I like it too:

```
$ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
$ docker pull localhost:5000/icecrime/busybox:latest
latest: Pulling from localhost:5000/icecrime/busybox

511136ea3c5a: Pull complete
df7546f9f060: Pull complete
ea13149945cb: Pull complete
4986bf8c1536: Pull complete
localhost:5000/icecrime/busybox:latest: The image you are pulling has been verified.
This is a tech preview, don't rely on it for security yet.

Status: Downloaded newer image for localhost:5000/icecrime/busybox:latest
```

**Comment by icecrime** [comment]
Once small nitpick: I'd replace the newline by a ` - ` to make it clear we're referring to the `"verified"` part, and not to the whole image pulling thing :p

Otherwise LGTM!

**Comment by jfrazelle** [comment]
Updated!

**Comment by tiborvass** [comment]
LGTM

File: graph/pull.go [hunk 5]
```diff
@@ -553,6 +554,8 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
 
 	}
 
+	out.Write(sf.FormatStatus(repoInfo.CanonicalName+":"+tag, "The image you are pulling has been verified - This is a tech preview, don't rely on it for security yet."))
```
Comment (fredlf) [inline graph/pull.go]: Arnaud asked me to look at this and I'm just getting to it now. 
I would suggest the following: "The image you are pulling has been verified. *Important:* image verification is a tech preview feature and should not be relied on to provide security."

Comment (jfrazelle) [inline graph/pull.go]: Ah ok that @fredlf I will ping you in a followup PR sorry we didn't wait.

**Pull request merged**